src/main.rs

```diff
@@ -43,312 +44,301 @@ mod rpc;
 mod textline;
 mod xi_thread;
 
-use std::any::Any;
-use std::cell::{RefCell, RefMut};
-use std::rc::Rc;
 use std::sync::{Arc, Mutex};
+use std::collections::HashMap;
 
 use winapi::shared::windef::*;
 
 use serde_json::Value;
+use serde_json::json;
 
 use edit_view::EditView;
 use menus::MenuEntries;
 use rpc::{Core, Handler};
-use xi_win_shell::util::Error;
 use dialog::{get_open_file_dialog_path, get_save_file_dialog_path};
 use xi_thread::start_xi_thread;
 
-use xi_win_shell::paint::PaintCtx;
-use xi_win_shell::win_main::{self, RunLoopHandle};
-use xi_win_shell::window::{Cursor, IdleHandle, MouseEvent, WindowBuilder,
-    WindowHandle, WinHandler};
+use xi_win_shell::win_main::{self};
+use xi_win_shell::window::{Cursor, IdleHandle, WindowBuilder};
 
-struct MainWinState {
-    edit_view: EditView,
+use xi_win_ui::{UiMain, UiState};
+use xi_win_ui::Id;
+
+use edit_view::EditViewCommands;
+
+type ViewId = String;
+
+#[derive(Clone)]
+struct ViewState {
+    id: Id,
+    filename: Option<String>,
+}
+
+#[derive(Clone)]
+struct AppState {
+    focused: ViewId,
+    views: HashMap<ViewId, ViewState>,
 }
 
-impl MainWinState {
-    fn new() -> MainWinState {
-        MainWinState {
-            edit_view: EditView::new(),
+impl AppState {
+    fn new() -> AppState {
+        AppState {
+            focused: Default::default(),
+            views: HashMap::new(),
         }
     }
 
-    fn render(&mut self, p: &mut PaintCtx) {
-        self.edit_view.render(p);
+    fn get_focused_viewstate(&mut self) -> &mut ViewState {
+        if let Some(state) = self.views.get_mut(&self.focused) {
+            return state
+        } else {
+            panic!("Getting viewstate failed.\nFocused: {}\n", &self.focused)
+        }
     }
 }
 
-struct MainWinHandler {
-    win: Rc<MainWin>,
-}
-
-// Maybe combine all this, put as a single item inside a RefCell.
-pub struct MainWin {
-    core: RefCell<Core>,
-    handle: RefCell<WindowHandle>,
-    state: RefCell<MainWinState>,
+#[derive(Clone)]
+struct App {
+    core: Arc<Mutex<Core>>,
+    state: Arc<Mutex<AppState>>,
+    handle: Arc<Mutex<IdleHandle>>,
```

Comment:
The `IdleHandle` should probably be in the ViewState, as if there are multiple windows you want the IdleHandle for the window holding the view.